woocommerce/woocommerce
# Add monorepo util to notify slack, add improvements to calling utils and type clean up.

### Changes proposed in this Pull Request:

1. Add a new command that allows you to send slack messages to a channel via an oauth token.
2. Remove bin/run code, improve program running code to not error in some situations
3. Remove some types we don't need to use thanks to `extra-typings` package
4. Update some deps, mainly to fix types and get latest commander.

Closes https://github.com/woocommerce/woocommerce/pull/38185



### How to test the changes in this Pull Request:

1. I think the nicest way to test this command without interfering with a real Slack instance is to [create a new slack workspace](https://slack.com/get-started#/createnew).
2. Now log in to the workspace and [create an app](https://api.slack.com/authentication/basics#creating).
3. In the app settings you need to [give it permissions](https://api.slack.com/authentication/basics#scopes) to send messages to generate a token. `chat:write` should be sufficient.
4. Once you give the scopes, save the [auth token](https://api.slack.com/authentication/basics#getting-your-authentication-token) generated in your password manager of choice.
5. Now checkout this branch
6. When you checkout the branch `postinstall` should have run and so the code should be built and ready to go, but if you want to do some debugging or adjust code to try something make sure to run `pnpm build` in `tools/monorepo-utils` to see changes reflected
7. Now from project root try some different commands: (`token` should be the oauth token you generated when creating your app)

~~7a. I adjusted how commands are run, so you should be able to see the help output but not have it exit(1) like it used to. e.g. to see the main help run `pnpm utils`. It will list available commands and there will be no `ELIFECYCLE` error.
7b. Now send a test message to your slack you created. `pnpm utils slack "token" "general" "Hello World, this is my message"`
7c. Now try send a message to non-existent channel. `pnpm utils slack "token" "channel_does_not_exist" "Uh oh!"`
7d. Note that with `7c` the command should exit with failure, but if you want the command to carry on even if the message was not sent you can try: `pnpm utils slack "token" "channel_does_not_exist" "Uh oh!" --dontfail`. It will show the error but it will not exit with failure. This might be useful if you don't want slack failure to break your CI action.~~

## Updated Test Instructions for Changes:

7a. Test help does not error, run at root: `pnpm utils`
7b. Send a message to single channel: `pnpm utils slack "token" "message" general`
7c. Send to multiple channels: `pnpm utils slack "token" "message" general random`
7d. Try send to multiple channels, fail on first: `pnpm utils slack "token" "message" not_exist general`
7e. Try to send to multiple, never fail (will send to all valid channels): `pnpm utils slack "token" "message" not_exist general --dont-fail`
